okhttp/src/main/java/okhttp3/MultipartBody.kt

```diff
@@ -0,0 +1,302 @@
+   * double-duty to make sure the counting and content are consistent, particularly when it comes
+   * to awkward operations like measuring the encoded length of header strings, or the
+   * length-in-digits of an encoded integer.
+   */
+  @Throws(IOException::class)
+  private fun writeOrCountBytes(
+    sink: BufferedSink?, countBytes: Boolean
+  ): Long {
+    var sink = sink
+    var byteCount = 0L
+
+    var byteCountBuffer: Buffer? = null
+    if (countBytes) {
+      byteCountBuffer = Buffer()
+      sink = byteCountBuffer
+    }
+
+    for (p in 0 until parts.size) {
+      val part = parts[p]
+      val headers = part.headers
+      val body = part.body
+
+      sink!!.write(DASHDASH)
+      sink.write(boundary)
+      sink.write(CRLF)
+
+      if (headers != null) {
+        for (h in 0 until headers.size()) {
+          sink.writeUtf8(headers.name(h))
+              .write(COLONSPACE)
+              .writeUtf8(headers.value(h))
+              .write(CRLF)
+        }
+      }
+
+      val contentType = body.contentType()
+      if (contentType != null) {
+        sink.writeUtf8("Content-Type: ")
+            .writeUtf8(contentType.toString())
+            .write(CRLF)
+      }
+
+      val contentLength = body.contentLength()
+      if (contentLength != -1L) {
+        sink.writeUtf8("Content-Length: ")
+            .writeDecimalLong(contentLength)
+            .write(CRLF)
+      } else if (countBytes) {
+        // We can't measure the body's size without the sizes of its components.
+        byteCountBuffer!!.clear()
+        return -1L
+      }
+
+      sink.write(CRLF)
+
+      if (countBytes) {
+        byteCount += contentLength
+      } else {
+        body.writeTo(sink)
+      }
+
+      sink.write(CRLF)
+    }
+
+    sink!!.write(DASHDASH)
+    sink.write(boundary)
+    sink.write(DASHDASH)
+    sink.write(CRLF)
+
+    if (countBytes) {
+      byteCount += byteCountBuffer!!.size
+      byteCountBuffer.clear()
+    }
+
+    return byteCount
+  }
+
+  class Part private constructor(internal val headers: Headers?, internal val body: RequestBody) {
+
+    fun headers(): Headers? = headers
+
+    fun body(): RequestBody = body
+
+    companion object {
+      @JvmStatic
+      fun create(body: RequestBody): Part = create(null, body)
+
+      @JvmStatic
+      fun create(headers: Headers?, body: RequestBody): Part {
+        require(headers?.get("Content-Type") == null) { "Unexpected header: Content-Type" }
+        require(headers?.get("Content-Length") == null) { "Unexpected header: Content-Length" }
+        return Part(headers, body)
+      }
+
+      @JvmStatic
+      fun createFormData(name: String, value: String): Part =
+          createFormData(name, null, RequestBody.create(null, value))
+
+      @JvmStatic
+      fun createFormData(name: String, filename: String?, body: RequestBody): Part {
+        val disposition = StringBuilder("form-data; name=")
```

Comment:
Could be `buildString`, or we can clean them all up later

Reply:
Later!